drivers/serial/eusart_gecko.c

```diff
@@ -0,0 +1,532 @@
+	advancedSettings.hwFlowControl = eusart_gecko_cfg2ll_hwctrl(uart_cfg->flow_ctrl);
+	eusartInit.advancedSettings = &advancedSettings;
+
+	EUSART_UartInitHf(config->eusart, &eusartInit);
+
+	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	config->irq_config_func(dev);
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int eusart_gecko_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	__maybe_unused const struct eusart_gecko_config *config = dev->config;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+#ifdef EUSART_STATUS_TXIDLE
+		/* Wait for TX FIFO to flush before suspending */
+		while (!(EUSART_StatusGet(config->eusart) & EUSART_STATUS_TXIDLE)) {
+		}
+#endif
+		break;
+
+	case PM_DEVICE_ACTION_RESUME:
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+#endif
+
+static DEVICE_API(uart, eusart_gecko_driver_api) = {
+	.poll_in = eusart_gecko_poll_in,
+	.poll_out = eusart_gecko_poll_out,
+	.err_check = eusart_gecko_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = eusart_gecko_fifo_fill,
+	.fifo_read = eusart_gecko_fifo_read,
+	.irq_tx_enable = eusart_gecko_irq_tx_enable,
+	.irq_tx_disable = eusart_gecko_irq_tx_disable,
+	.irq_tx_complete = eusart_gecko_irq_tx_complete,
+	.irq_tx_ready = eusart_gecko_irq_tx_ready,
+	.irq_rx_enable = eusart_gecko_irq_rx_enable,
+	.irq_rx_disable = eusart_gecko_irq_rx_disable,
+	.irq_rx_ready = eusart_gecko_irq_rx_ready,
+	.irq_err_enable = eusart_gecko_irq_err_enable,
+	.irq_err_disable = eusart_gecko_irq_err_disable,
+	.irq_is_pending = eusart_gecko_irq_is_pending,
+	.irq_update = eusart_gecko_irq_update,
+	.irq_callback_set = eusart_gecko_irq_callback_set,
+#endif
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+#define GECKO_EUSART_IRQ_HANDLER_DECL(idx)				       \
+	static void eusart_gecko_config_func_##idx(const struct device *dev)
+#define GECKO_EUSART_IRQ_HANDLER_FUNC(idx)				       \
+	.irq_config_func = eusart_gecko_config_func_##idx,
+#define GECKO_EUSART_IRQ_HANDLER(idx)					       \
+	static void eusart_gecko_config_func_##idx(const struct device *dev)   \
+	{								       \
+		IRQ_CONNECT(DT_INST_IRQ_BY_NAME(idx, rx, irq),		       \
+			    DT_INST_IRQ_BY_NAME(idx, rx, priority),	       \
+			    eusart_gecko_isr, DEVICE_DT_INST_GET(idx), 0);     \
+		IRQ_CONNECT(DT_INST_IRQ_BY_NAME(idx, tx, irq),		       \
+			    DT_INST_IRQ_BY_NAME(idx, tx, priority),	       \
+			    eusart_gecko_isr, DEVICE_DT_INST_GET(idx), 0);     \
+									       \
+		irq_enable(DT_INST_IRQ_BY_NAME(idx, rx, irq));		       \
+		irq_enable(DT_INST_IRQ_BY_NAME(idx, tx, irq));		       \
+	}
+#else
+#define GECKO_EUSART_IRQ_HANDLER_DECL(idx)
+#define GECKO_EUSART_IRQ_HANDLER_FUNC(idx)
+#define GECKO_EUSART_IRQ_HANDLER(idx)
+#endif
+
+#define GECKO_EUSART_INIT(idx)							\
+GECKO_EUSART_IRQ_HANDLER_DECL(idx);						\
```

Comment:
I suggest to use `GECKO_EUSART_IRQ_HANDLER_FUNC()` here and to drop `GECKO_EUSART_IRQ_HANDLER_DECL()`.

Reply:
I tend to agree but I'm wondering why this is done in so many other places...